include/nlohmann/detail/json_pointer.hpp

```diff
@@ -841,11 +841,10 @@ class json_pointer
 
     @exceptionsafety No-throw guarantee: this function never throws exceptions.
     */
-    friend bool operator==(json_pointer const& lhs,
-                           json_pointer const& rhs) noexcept
-    {
-        return lhs.reference_tokens == rhs.reference_tokens;
-    }
+    template<typename RefStringTypeLhs, typename RefStringTypeRhs>
+    // NOLINTNEXTLINE(readability-redundant-declaration)
+    friend bool operator==(::nlohmann::json_pointer<RefStringTypeLhs> const&,
+                           ::nlohmann::json_pointer<RefStringTypeRhs> const&) noexcept;
```

Comment:
Do these need the `::nlohmann::` prefix, and why do they need to be moved out of line?

Reply:
> Do these need the `::nlohmann::` prefix, and
Not that I'm aware. It's just a preference given `using json_pointer = ...` in `basic_json` and me wanting to draw a clear distinction between the actual type and some typedef. I can change that if you think it makes it more readable or something?

> why do they need to be moved out of line?
Because to not do so would result in ODR violations. (Since both the class and the functions are templated and the same function could be instantiated for two or more classes.)

Reply:
Seems like since it's inside the `json_pointer` class in the `nlohmann` namespace, this seems to be adding unnecessary verbosity.  To be honest, If it was already `::nlohmann::json_pointer`, I probably wouldn't have even noticed, just noticed because it's a change.  Is this related to moving it out of the class body?

> Because to not do so would result in ODR violations.

Could you provide more details on how having an inline friend function body inside a class definition is any different than a forward friend declaration and an inline function inside the containing namespace?  I'm not familiar with that issue.  Since it's inline, and the exact same token sequence for each, there shouldn't be any issue as far as I know.

Reply:
> Seems like since it's inside the `json_pointer` class in the `nlohmann` namespace, this seems to be adding unnecessary verbosity. To be honest, If it was already `::nlohmann::json_pointer`, I probably wouldn't have even noticed, just noticed because it's a change. Is this related to moving it out of the class body?

I mean it's *because* it's inside the class and there is this pattern in this library to have typedefs that can cause confusion. This is not the case here specifically. I can change it to be less verbose.

> 
> > Because to not do so would result in ODR violations.
> 
> Could you provide more details on how having an inline friend function body inside a class definition is any different than a forward friend declaration and an inline function inside the containing namespace? I'm not familiar with that issue. Since it's inline, and the exact same token sequence for each, there shouldn't be any issue as far as I know.

Sure. You can read all about the gory details in this language defect report:  http://www.open-std.org/jtc1/sc22/wg21/docs/cwg_defects.html#2174

Reply:
Done.

Reply:
Thanks.  I had a vague recollection of that, and I think the issue has to do with template partial specializations, but I suppose it doesn't hurt in this case.

Reply:
I ported this from my string_view3 branch, so I wrote it in early March. I do recall the compiler barking at me. It's definitely required.

Reply:
Thanks.